### PR TITLE
Add auto complete suggestions in eval text field

### PR DESCRIPTION
**Pull request by jankuss**
This PR adds the currently hovered auto complete item as a suggestion at the end of the eval text field.

## Before
image

## After

https://user-images.githubusercontent.com/1659532/166514287-e3123d72-c810-429d-9dd7-147bf1cda145.mov

This should improve the overall UX of the eval text field.

This fixes bullet point number 2 in #3095 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/devtools_app/lib/src/ui/search.dart [hunk 1]
```diff
@@ -297,20 +297,24 @@ class AutoCompleteState extends State<AutoComplete> with AutoDisposeMixin {
...
     final areaHeight = offset.dy;
+
     final maxAreaForPopup = areaHeight - tileEntryHeight;
     // TODO(terry): Scrolling doesn't work so max popup height is also total
     //              matches to use.
-    topMatchesLimit = min(
-      defaultTopMatchesLimit,
-      (maxAreaForPopup / tileEntryHeight) - 1, // zero based.
-    ).truncate();
+    topMatchesLimit = tileEntryHeight > 0
```
Comment (elliette) [inline packages/devtools_app/lib/src/ui/search.dart:305]: What is this check for (when would the tileEntryHeight be 0?)

File: packages/devtools_app/lib/src/screens/debugger/evaluate.dart [hunk 2]
```diff
@@ -176,6 +237,11 @@ class _ExpressionEvalFieldState extends State<ExpressionEvalField>
                   ),
                 );
               },
+              // Disable ligatures, so the suggestions of the auto complete work correcly.
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/debugger/evaluate.dart:240]: What happens with ligatures enabled?

**Comment by elliette** [comment]
This looks great, thank you! 🎉  🎉 

Would it be possible to update so that if the cursor is at the end of the text field and a user hits the right arrow key, the suggestion is entered?

**Comment by kenzieschmoll** [comment]
> Would it be possible to update so that if the cursor is at the end of the text field and a user hits the right arrow key, the suggestion is entered?

Implementing the same for when a user hits 'tab' would also be great

File: packages/devtools_app/lib/src/screens/debugger/evaluate.dart (hunk 2 above)
Comment (jankuss) [reply packages/devtools_app/lib/src/screens/debugger/evaluate.dart:240]: This is what happens:

https://user-images.githubusercontent.com/1659532/166835194-ca3bef40-74be-4143-a39c-9d8c019ca49b.mov

T‌he ligature in this example is the `fi` of in the word `profile`. I'm not quite sure why this behaviour happens though, maybe do you have an idea?

T‌o build the content within the `TextField`, we use this code:
image

We have two `TextSpan`, the first displays the text `prof` (which is the actual typed text). The other `TextSpan` will display `ile`, which is the suggestion. The only reasonable cause I can come up with is that something goes wrong because the ligature is split between two different `TextSpan`.

I tried placing a "Zero Width Non-Joiner" (https://unicode-explorer.com/c/200C) character to prevent this, but this will break the caret position.

That's why the easiest solution to this is to disable ligatures for the text in the TextField.

File: packages/devtools_app/lib/src/ui/search.dart (hunk 1 above)
Comment (jankuss) [reply packages/devtools_app/lib/src/ui/search.dart:305]: The `tileEntryHeight` will be zero, when there are no items in the auto complete popup. Interestingly, the logic im guarding with that check works fine in the actual application, it will just fail in tests 🤷 

Thinking about it, since we don't display any auto complete items in the popup when there are no matches, we could just `return` early and skip the unnecessary calculation which is causing issues.

Basically a early return along the lines of:
```
if (tileContents.isEmpty) return const SizedBox();
```

**Comment by jankuss** [comment]
> This looks great, thank you! 🎉 🎉
> 
> Would it be possible to update so that if the cursor is at the end of the text field and a user hits the right arrow key, the suggestion is entered?

Will add it 👍 

> > Would it be possible to update so that if the cursor is at the end of the text field and a user hits the right arrow key, the suggestion is entered?
> 
> Implementing the same for when a user hits 'tab' would also be great

This was already possible before the PR

File: packages/devtools_app/lib/src/screens/debugger/evaluate.dart (hunk 2 above)
Comment (elliette) [reply packages/devtools_app/lib/src/screens/debugger/evaluate.dart:239]: Oh weird! Thanks for the explanation. Fine to keep as is with ligatures disabled

File: packages/devtools_app/lib/src/ui/search.dart [hunk 3]
```diff
@@ -285,6 +285,10 @@ class AutoCompleteState extends State<AutoComplete> with AutoDisposeMixin {
         )
         .toList();
 
+    // When there are no tiles present, we don't need to display the
+    // auto correct list.
+    if (tileContents.isEmpty) return const SizedBox.shrink();
```
Comment (elliette) [inline packages/devtools_app/lib/src/ui/search.dart:290]: This seems like a good solution, thanks!

**Review by elliette — APPROVED** [review]
Thank you for adding this!

File: packages/devtools_app/test/eval_field_test.dart [hunk 4]
```diff
@@ -0,0 +1,251 @@
+import 'package:devtools_app/devtools_app.dart';
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/eval_field_test.dart:1]: this file needs a copy right. You can copy from another test file and paste here, just make sure the year is 2022. Also, please move this test to the `test/debugger/` directory

File: packages/devtools_app/lib/src/screens/debugger/evaluate.dart [hunk 5]
```diff
@@ -65,12 +74,63 @@ class _ExpressionEvalFieldState extends State<ExpressionEvalField>
...
+    final editingParts = _getCurrentEditingParts();
+
+    if (!_isRealVariableNameOrField(editingParts)) {
+      return;
+    }
+
+    _autoCompleteController.updateCurrentSuggestion(_activeWord);
+  }
+
+  EditingParts _getCurrentEditingParts() {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/debugger/evaluate.dart:103]: nit: avoid starting method names with "get". Let's replace with `_currentEditingParts`

File: packages/devtools_app/test/eval_field_test.dart (hunk 4 above)
Comment (jankuss) [reply packages/devtools_app/test/eval_field_test.dart:5]: There is no `test/debugger/` directory, or do you mean that the test file should be prefixed with `debugger_`?

Comment (jankuss) [reply packages/devtools_app/test/eval_field_test.dart]: I see - Added the copyright and moved the file

**Review by kenzieschmoll — APPROVED** [review]
